monitor/guard.go

```diff
@@ -1,151 +1,185 @@
-		case *mainchain.CelerLedgerIntendWithdraw:
-			m.guardIntendWithdrawChannel(e)
+		keys = append(keys, iterator.Key())
+		vals = append(vals, iterator.Value())
+	}
+	iterator.Close()
+	m.dbLock.RUnlock()
+
+	for i, key := range keys {
+		event := NewEventFromBytes(vals[i])
+		if !event.Processing {
+			log.Infoln("Process guard event", event.Name)
+			var submitted bool
+			switch e := event.ParseEvent(m.ethClient).(type) {
+			case *mainchain.CelerLedgerIntendSettle:
+				e.Raw = event.Log
+				submitted, err = m.guardIntendSettle(e)
+			case *mainchain.CelerLedgerIntendWithdraw:
+				e.Raw = event.Log
+				submitted, err = m.guardIntendWithdrawChannel(e)
+			}
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			if submitted {
+				m.dbLock.Lock()
+				v, err2 := m.db.Get(key)
+				if err2 != nil {
+					log.Errorln("db Get err:", err)
+					m.dbLock.Unlock()
+					continue
+				}
+				e := NewEventFromBytes(v)
+				if !e.Processing {
+					e.Processing = true
+					err = m.db.Set(key, e.MustMarshal())
+					if err != nil {
+						log.Errorln("db Set err", err)
+					}
+				}
+				m.dbLock.Unlock()
+			}
 		}
 	}
 }
 
-func (m *Monitor) guardIntendSettle(intendSettle *mainchain.CelerLedgerIntendSettle) {
+func (m *Monitor) guardIntendSettle(intendSettle *mainchain.CelerLedgerIntendSettle) (bool, error) {
 	log.Infof("Guard IntendSettle %x, tx hash %x", intendSettle.ChannelId, intendSettle.Raw.TxHash)
 	requests := m.getRequests(intendSettle.ChannelId)
-	for _, request := range requests {
-		m.guardRequest(request, intendSettle.Raw, IntendSettle)
+	if len(requests) > 0 {
+		return m.guardRequest(requests, intendSettle.Raw, IntendSettle)
+	} else {
+		err := m.dbDelete(GetGuardKey(intendSettle.Raw))
+		return false, err
 	}
 }
 
-func (m *Monitor) guardIntendWithdrawChannel(intendWithdrawChannel *mainchain.CelerLedgerIntendWithdraw) {
+func (m *Monitor) guardIntendWithdrawChannel(intendWithdrawChannel *mainchain.CelerLedgerIntendWithdraw) (bool, error) {
 	log.Infof("Guard intendWithdrawChannel %x, tx hash %x", intendWithdrawChannel.ChannelId, intendWithdrawChannel.Raw.TxHash)
 	requests := m.getRequests(intendWithdrawChannel.ChannelId)
-	for _, request := range requests {
-		m.guardRequest(request, intendWithdrawChannel.Raw, IntendWithdrawChannel)
+	if len(requests) > 0 {
+		return m.guardRequest(requests, intendWithdrawChannel.Raw, IntendWithdrawChannel)
+	} else {
+		err := m.dbDelete(GetGuardKey(intendWithdrawChannel.Raw))
+		return false, err
 	}
 }
 
-func (m *Monitor) getRequests(cid mainchain.CidType) (requests []subscribe.Request) {
-	addresses, seqNums, err := m.ethClient.Ledger.GetStateSeqNumMap(
-		&bind.CallOpts{BlockNumber: sdk.NewIntFromUint64(m.secureBlkNum).BigInt()},
-		cid)
-	if err != nil {
-		log.Errorln("Query StateSeqNumMap err", err)
-		return
+func (m *Monitor) guardRequest(requests []*subscribe.Request, rawLog ethtypes.Log, eventName EventName) (bool, error) {
+	if len(requests) != 1 && len(requests) != 2 {
```

Comment:
len(requests) <= 2

Reply:
I wanted to make len=0 also invalid, though the current callers already excluded the cases